servlet/src/main/java/com/linecorp/armeria/server/servlet/ServletPrintWriter.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.servlet;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.util.Formatter;
+import java.util.Locale;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpData;
+
+/**
+ *  Printing flow.
+ */
+final class ServletPrintWriter extends PrintWriter {
+    private final String lineSeparator = System.lineSeparator();
+    private final DefaultServletHttpResponse response;
+
+    private boolean error;
+
+    ServletPrintWriter(DefaultServletHttpResponse response, OutputStream out) {
+        super(out);
+        requireNonNull(response, "response");
+        this.response = response;
+    }
+
+    @Override
+    public void flush() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public void close() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public boolean checkError() {
+        return error;
+    }
+
+    @Override
+    protected void setError() {
+        error = true;
+    }
+
+    @Override
+    protected void clearError() {
+        error = false;
+    }
+
+    @Override
+    public void write(int c) {
+        write(String.valueOf(c));
+    }
+
+    @Override
+    public void write(char[] buf, int off, int len) {
+        requireNonNull(buf, "buf");
+        checkArgument(off >= 0, "off: %s (expected: >= 0)", off);
+        checkArgument(len >= 0, "len: %s (expected: >= 0)", len);
+        write(String.valueOf(buf, off, len));
+    }
+
+    @Override
+    public void write(char[] buf) {
+        requireNonNull(buf, "buf");
+        write(String.valueOf(buf));
+    }
+
+    @Override
+    public void write(String s, int off, int len) {
+        requireNonNull(s, "s");
+        checkArgument(off >= 0, "off: %s (expected: >= 0)", off);
+        checkArgument(len >= 0, "len: %s (expected: >= 0)", len);
+        final String writeStr;
+        if (off == 0 && s.length() == len) {
+            writeStr = s;
+        } else {
+            writeStr = s.substring(off, off + len);
+        }
+        response.write(HttpData.ofUtf8(writeStr));
```

Comment:
ditto `write` can only happen once?

Reply:
Do you mean we should check state `isWrote` ?